---
title: "Machine learning for biological contexts"
output:
  html_notebook:
    toc: true
    toc_float: true
date: 2025
---

_This material has been adapted from [ALSF CCDL training materials](https://github.com/AlexsLemonade/training-modules/tree/master/machine-learning)._

## Objectives

* Illustrate how to use the `PLIER` method for unsupervised machine learning for human transcriptomics data
* Demonstrate how to create a heatmap with the `ComplexHeatmap` package
* Introduce the concept of tidy data
* Briefly introduce customizing `ggplot2` plots (but there will be more on `ggplot2` in a later lesson).

## Background

As we've seen in the course so far, we can explore data with unsupervised machine learning approaches like clustering or PCA.
Often, these methods can work with any generic dataset.
In this notebook, we'll introduce a machine learning technique that is specifically for gene expression data.

The dataset we're using comes from the [OpenPBTA project](https://github.com/AlexsLemonade/OpenPBTA-analysis).
We'll be using medulloblastoma data only.

## Set Up

```{r}
# Bit o' data wranglin' expected
library(tidyverse)

# Heatmap
library(ComplexHeatmap)
```

Make sure `PLIER` is installed.

```{r}
if (!("remotes" %in% installed.packages())) {
  install.packages("remotes")
}

if (!("PLIER" %in% installed.packages())) {
  # Install PLIER from GitHub
  remotes::install_github("wgmao/PLIER@v0.1.6")
}

# Load Pathway-Level Information ExtractoR
library(PLIER)
```

## Read in and set up data

## Pathway-Level Information ExtractoR (PLIER)

In this notebook, we'll use a method called Pathway-Level Information Extractor (PLIER) ([Mao *et al.* (2019)](https://doi.org/10.1038/s41592-019-0456-1)).

We like PLIER for a few reasons:

* It is a matrix factorization approach.
  That means we can get a low-dimensional representation of our data.
  Specifically, PLIER learns correlated patterns of expression in our data or latent variables (LVs).
  Here, a latent variable is an "eigengene-like" combination of genes' expression.
  (It's called _latent_ because it's not directly measured, but instead inferred from the individual gene measurements.)
* It includes penalties such that _some_ of the LVs will align with gene sets that we give it, so it's excellent for biological discovery.
* The authors demonstrated that it performs favorably with regard to estimating proportion of immune cells in a sample as compared to other methods.
* Because not _all_ LVs align with the gene sets we input, some of them capture unwanted technical variation.
  In our experience with the method, it does this quite well.

PLIER is similar to other pathway analysis methods that you may be familiar with in that it uses prior knowledge in the form of gene sets.
It produces output values that are on an individual sample level and does not require a two group comparison ahead of time like some pathway analysis methods.
However, PLIER is designed to align the LVs it constructs with the relevant input gene sets that the data supports, whereas other methods will use all gene sets you provide as input.

Here's an overview of the PLIER method from [Mao _et al._ (2019)](https://doi.org/10.1038/s41592-019-0456-1) (Figure 1).

![](diagrams/mao_nature_methods_fig1.png)

> **Fig. 1 | PLIER overview.**
PLIER is a matrix factorization approach that decomposes gene expression data into a product of a small number of LVs and their corresponding gene associations or loadings, while constraining the loadings to align with the most relevant automatically selected subset of prior knowledge. **a**, Given two inputs, the gene expression matrix _Y_ and the prior knowledge (represented as binary gene set membership in matrix _C_), the method returns the LVs (_B_), their loadings (_Z_), and an additional sparse matrix (_U_) that specifies which (if any) prior-information gene sets and pathways are used for each LV. The light gray area of _U_ indicates the large number of zero elements of the matrix. We apply our method to a whole-blood human gene expression dataset. **b**, The positive entries of the resulting _U_ matrix are visualized as a heat map, facilitating the identification of the correspondence between specific LVs and prior biological knowledge. As the absolute scale of the _U_ matrix is arbitrary, each column is normalized to a maximum of 1. **c**, We validate the LVs mapped to specific leukocyte cell types by comparing PLIER estimated relative cell-type proportions with direct measurements by mass cytometry. Dashed lines represent 0.05, 0.01, and 0.001 significance levels for Spearman rank correlation (one-tailed test). NK cell, natural killer cell.

### Read in and explore the model

We've prepared the model ahead of time to save time during the course.
You can see what steps we took to complete model training [here](https://github.com/jaclyn-taroni/2025-mdibl-fair/tree/main/setup/machine-learning/scripts/02-train-plier-model.R) and, more generally, how we setup the module [here](https://github.com/jaclyn-taroni/2025-mdibl-fair/tree/main/setup/machine-learning/).

```{r}
# The file containing the PLIER::PLIER() output is saved in the results
# directory
plier_file <- file.path(
  "results",
  "plier",
  "medulloblastoma_plier_model.rds"
)

# Read in the RDS file that contains the PLIER::PLIER() output
plier_results <- read_rds(plier_file)
```

What does the output of model training look like?

```{r view_plier, live = TRUE, eval = FALSE}
View(plier_results)
```

The _U_ matrix tells us about how the latent variables learned by the model relate to the pathways we used as input.
`plotU()` is a special function to display the _U_ matrix.

```{r plot_u}
PLIER::plotU(plier_results,
  fontsize_row = 6
)
```

`summary()` of a `PLIER` results object returns the FDR and AUC values for input pathway to latent variable relationships.

```{r lv_summary}
plier_results$summary %>%
  filter(FDR < 0.05) %>%
  arrange(FDR)
```

The _B_ matrix contains the latent variable values for each sample.

```{r dim_b}
dim(plier_results$B)
```

Let's take a peek at the matrix itself.

```{r b_preview, live = TRUE}
plier_results$B[1:5, 1:5]
```

The _Z_ matrix contains the gene loadings (how genes combine to get _B_).

```{r dim_z}
dim(plier_results$Z)
```

We can use _Z_ to tell us which genes contribute to individual LVs by accessing the column corresponding to that LV.
We'll use 20 below, but you can change the number to suit your purposes!

```{r lv_loadings}
head(
  sort(plier_results$Z[, 20],
    decreasing = TRUE
  ),
  n = 25
)
```

## Biological exploration with PLIER

### LVs associated with pathways

For biological discovery, we are often most interested in the latent variables that have some kind of association with an input gene set or pathway.
We can use the FDR values in the summary data frame to filter to only the latent variables with a significant association (and their associated gene sets).

```{r filter_fdr, live = TRUE}
# Filter to LV-pathway relationships with FDR < 0.05
sig_summary_df <- plier_results$summary %>%
  dplyr::filter(FDR < 0.05)
sig_summary_df
```

```{r sig_index}
# We only want a single instance of each LV index
sig_index <- as.integer(unique(sig_summary_df$`LV index`))
```

```{r sig_b_wide, live = TRUE}
# Get the LV by sample matrix from the PLIER results and subset it to only those
# LVs with an FDR < 0.05 (at least one pathway)
b_matrix <- plier_results$B
sig_b_matrix <- b_matrix[sig_index, ]
```

#### Heatmap

Let's make a heatmap of the latent variable values for the variables that are significantly associated with an input pathway.

We can make one pretty easily using the `ComplexHeatmap` package.

```{r}
Heatmap(sig_b_matrix)
```

I'm not sure that's so useful on its own, so let's make some improvements!

Medulloblastoma has molecular subtypes, and we have molecular subtype labels for these samples.
We can use this information to annotate our heatmap, but first, we need to read it in!

```{r}
# Read in metadata
histologies_df <- read_tsv(
  file.path(
    "data",
    "metadata",
    "pbta-histologies-medulloblastoma-rnaseq.tsv"
  )
)

# Create a data frame that only has the biospecimen identifiers and the
# molecular subtype labels
subtype_df <- histologies_df %>%
  select(
    Kids_First_Biospecimen_ID,
    molecular_subtype
  )
```
Heatmap annotations require the sample identifiers to be the rownames, so let's set that up.

```{r}
annotation_df <- subtype_df |>
  tibble::column_to_rownames("Kids_First_Biospecimen_ID") |>
  as.data.frame()
```

Now we're ready to make a heatmap annotation using a palette that is color vision deficiency friendly.

```{r}
# Get a vector of hex codes for the Okabe-Ito palette
okabe_ito_palette <- unname(palette.colors(palette = "Okabe-Ito"))

# Create a sample HeatmapAnnotation
sample_annotation <- HeatmapAnnotation(
  # Sample to molecular subtype mapping
  df = annotation_df,
  # Colors for the annotation
  col = list(molecular_subtype = c(
    "MB, Group3" = okabe_ito_palette[1],
    "MB, Group4" = okabe_ito_palette[2],
    "MB, SHH" = okabe_ito_palette[3],
    "MB, To be classified" = okabe_ito_palette[4],
    "MB, WNT" = okabe_ito_palette[5]
  )),
  # Make the label for the annotation look a bit nicer than the column name
  # would
  annotation_label = "Molecular Subtype"
)
```

Check that the order of samples is the same in the annotation and the matrix being annotated.

```{r}
identical(rownames(annotation_df), colnames(sig_b_matrix))
```

Now we can make a heatmap object with some adjustments explained in the inline comments.
We'll plot it in the next chunk where we adjust the legend position.

```{r}
heatmap_object <- Heatmap(sig_b_matrix,
  # Add molecular subtype annotation
  top_annotation = sample_annotation,
  # The sample names were hard to read
  show_column_names = FALSE,
  # Make the row names a bit smaller
  row_names_gp = gpar(fontsize = 6),
  # Let's add some space between the cells
  rect_gp = gpar(col = "white", lwd = 0.25),
  # Make the heatmap legend horizontal instead of
  # Vertical
  heatmap_legend_param = list(direction = "horizontal")
)
```

To adjust the legend positions, we can use `heatmap_legend_side` and `annotation_legend_side` with `draw()`.

```{r}
draw(heatmap_object,
  # Put heatmap legend below the heatmap
  heatmap_legend_side = "bottom",
  # Put the annotation legend below the heatmap
  annotation_legend_side = "bottom"
)
```

Let's save this heatmap as a PNG.

It can be helpful to keep all the plots organized in the same folder, so let's set that up first.

```{r}
# Call the folder plots
plots_dir <- file.path("plots")
# Create it if it doesn't exist yet
dir.create(plots_dir, showWarnings = FALSE, recursive = TRUE)
```

Now, save as a PNG in the `plots` directory.

```{r}
heatmap_file <- file.path(
  plots_dir,
  "medulloblastoma_significant_lvs_heatmap.png"
)

# Plotting device
png(heatmap_file, width = 8, height = 5, units = "in", res = 300)
# Draw the heatmap and legends again
draw(heatmap_object,
  # Put heatmap legend below the heatmap
  heatmap_legend_side = "bottom",
  # Put the annotation legend below the heatmap
  annotation_legend_side = "bottom"
)
# Shut down device
dev.off()
```

### First, a note on tidy data

In order to use `ggplot2`, we'll need the data in "long" or "tidy" format.
`PLIER` outputs what we want to plot in what we call "wide" format.

Read more about tidy data [here](https://cran.r-project.org/web/packages/tidyr/vignettes/tidy-data.html).

To quote from Hadley Wickham's [R for Data Science](https://r4ds.had.co.nz/):

> There are three interrelated rules which make a dataset tidy:
>
> * Each variable must have its own column.
>
> * Each observation must have its own row.
>
> * Each value must have its own cell.

Let's look at a toy example.

```{r create_toy_wide}
set.seed(12345)
toy_df <- data.frame(
  cbind(
    c("GENEA", "GENEB", "GENEC"),
    matrix(rnorm(30), ncol = 10)
  )
)
colnames(toy_df) <- c("Gene", paste0("Sample", 1:10))
```

`toy_df` is now in "wide" format.

```{r toy_df}
toy_df
```

Let's get it into "long" format.

```{r make_toy_long}
toy_long_df <- tidyr::pivot_longer(toy_df,
  # The data is in every column except the one
  # named "Gene"
  cols = -Gene,
  # What will we call the column of the old df
  # column names?
  names_to = "Sample",
  # What will we call the column of values
  # from the old df?
  values_to = "Expression"
)

toy_long_df
```

Let's remove these toy examples from the workspace.

```{r remove_toy, live = TRUE}
rm(toy_df, toy_long_df)
```

### Tidy latent variables

Let's look at what format the LV values are in currently.

```{r wide_peek}
# First, create a data frame of and add a column with LV identifiers
sig_b_wide <- data.frame(sig_b_matrix) %>%
  tibble::rownames_to_column(var = "LV")

sig_b_wide
```

We want this in long format for plotting.
We'll use `tidyr::pivot_longer` to do this just like in the toy example above.

```{r pivot_longer_b, live = TRUE}
sig_b_df <- tidyr::pivot_longer(sig_b_wide,
  cols = starts_with("BS_"),
  names_to = "Kids_First_Biospecimen_ID",
  values_to = "LV_estimate"
)
head(sig_b_df)
```

Right now the `LV` column has values that contain two pieces of information: the LV index and the pathway that the LV has been named for.

*Remember, just because a LV is named for a single pathway, that doesn't mean that that is the only input pathway that is significantly associated with that latent variable - always check `summary`!*

Now let's add relevant metadata to the data frame so we can use that for plotting.

```{r add_metadata_b}
# Add the subtype labels to the LV estimates
sig_b_df <- inner_join(
  x = sig_b_df,
  y = subtype_df,
  by = "Kids_First_Biospecimen_ID"
)
```
## Plotting

We'll plot LV20; this is the latent variable that we looked at the loadings for in an earlier chunk.
You can try using a different LV if you would like!

```{r lv_to_plot}
# PLIER names certain latent variables based on their association with input
# gene sets
lv_to_plot <- rownames(plier_results$B)[20]
# For plotting, subset only to the rows corresponding to this latent variable
lv_plot_df <- sig_b_df %>%
  filter(LV == lv_to_plot)
```

### Boxplot and customization

Let's start by making a simple boxplot.

```{r lv_boxplot, live = TRUE}
# Make a boxplot where samples are grouped by molecular subtype
ggplot(
  lv_plot_df,
  aes(
    x = molecular_subtype,
    y = LV_estimate,
    group = molecular_subtype,
    color = molecular_subtype
  )
) +
  geom_boxplot()
```

It can often be helpful to visualize individual samples.

```{r lv_jitter, live = TRUE}
# Add individual points with geom_jitter()
ggplot(
  lv_plot_df,
  aes(
    x = molecular_subtype,
    y = LV_estimate,
    group = molecular_subtype,
    color = molecular_subtype
  )
) +
  geom_boxplot(outlier.shape = NA) +
  geom_jitter()
```

We're able to globally adjust the aesthetics of the jitter points.

```{r improve_jitter, live = TRUE}
# Improve the aesthetics of the points
ggplot(
  lv_plot_df,
  aes(
    x = molecular_subtype,
    y = LV_estimate,
    group = molecular_subtype,
    color = molecular_subtype
  )
) +
  geom_boxplot(outlier.shape = NA) +
  geom_jitter(width = 0.2, alpha = 0.5)
```

Add a built-in `ggplot2` theme.

```{r theme_bw, live = TRUE}
# Use @jaclyn-taroni's favorite theme :)
ggplot(
  lv_plot_df,
  aes(
    x = molecular_subtype,
    y = LV_estimate,
    group = molecular_subtype,
    color = molecular_subtype
  )
) +
  geom_boxplot(outlier.shape = NA) +
  geom_jitter(width = 0.2, alpha = 0.5) +
  theme_bw()
```

Use a different color palette.

```{r okabe_ito, live = TRUE}
# Add "Okabe-Ito" color scheme
ggplot(
  lv_plot_df,
  aes(
    x = molecular_subtype,
    y = LV_estimate,
    group = molecular_subtype,
    color = molecular_subtype
  )
) +
  geom_boxplot(outlier.shape = NA) +
  geom_jitter(width = 0.2, alpha = 0.5) +
  theme_bw() +
  scale_color_manual(values = unname(palette.colors(palette = "Okabe-Ito")))
```

Add a title to the plot.

```{r add_title, live = TRUE}
# Use labs() to add a title
ggplot(
  lv_plot_df,
  aes(
    x = molecular_subtype,
    y = LV_estimate,
    group = molecular_subtype,
    color = molecular_subtype
  )
) +
  geom_boxplot(outlier.shape = NA) +
  geom_jitter(width = 0.2, alpha = 0.5) +
  theme_bw() +
  scale_color_manual(values = unname(palette.colors(palette = "Okabe-Ito"))) +
  labs(title = lv_to_plot)
```

Center the title and make it bigger and in bold.

```{r center_title, live = TRUE}
# Use theme() to improve the way the title looks
ggplot(
  lv_plot_df,
  aes(
    x = molecular_subtype,
    y = LV_estimate,
    group = molecular_subtype,
    color = molecular_subtype
  )
) +
  geom_boxplot(outlier.shape = NA) +
  geom_jitter(width = 0.2, alpha = 0.5) +
  theme_bw() +
  scale_color_manual(values = unname(palette.colors(palette = "Okabe-Ito"))) +
  labs(title = lv_to_plot) +
  theme(plot.title = element_text(
    size = 15,
    face = "bold",
    hjust = 0.5
  ))
```

Use the next chunks to further customize your plot.
We might suggest starting with [the x- and y-axis labels](https://ggplot2.tidyverse.org/reference/labs.html).

```{r labs_help, eval = FALSE}
?labs
```

Try a new customization!

```{r adventure_time}
```

Save the last plot to a PNG file.
`ggplot2` has a function named `ggsave()` that we can use to do that.

```{r}
boxplot_file <- file.path(
  plots_dir,
  "medulloblastoma_plier_lv20_boxplot.png"
)
ggsave(boxplot_file, plot = last_plot())
```


## Session Info

```{r}
sessionInfo()
```

